### PR TITLE
Add function to run box models

### LINKED ISSUE #24: Create toy example using OceanBioME.jl 0D wrapper
We need an initial toy 0D model. We can rely on an existing wrapper by OceanBioME.jl initially (which relies on Oceananigans.jl). This can then be used for tests with Turing.jl and AD packages.

There is an initial version of the toy model here:
https://github.com/agate-model/Agate.jl/blob/dynamic_parsing/src/box_model/box_wrapper.jl

Many things to tidy:

`base_maximum_growth=0.69, maximum_grazing_rate= 2.1522`  should be an array
`PAR` definition should be outside of wrapper
` N = 7.0, P = 0.01, Z = 0.05` these initial conditions should be passed as an array into the wrapper
`if timeseries` only exports P currently

### PR DESCRIPTION
**Pull request by radka-j**
Closes #24 

## Overview

This PR adds a `run_box_model()` function which takes in an Agate generated model and passes that to `OceanBioME.BoxModel`. The user also specifies initial conditions of the simulation (and optionally the timestep and how long to run it for) and the function returns a NamedTuple of results for each tracer. 

The `examples/box_model.jl` script shows how this is used in practice. Once we have more complex models it should be possible to simply substitute them in place of the NPZD model here.

Similarly, I added an example of using `DifferentialEquations` to solve a box model. 

I also moved the NPZD parameter and tracer definitions to `examples/NPZD/` and now they are imported in the two new examples as well as inside tests so that we don't have to redefine them in multiple places. Our NPZD box model run is validated against OceanBioME NPZD box model (just a sanity check).

## For reviewers

Some questions:
- should the `run_box_model() function take in other input arguments?
- we use the same PAR function definition in a few places, should this be in the library somewhere?

### CODE REVIEW

**Comment by radka-j** [comment]
@nanophyto I just added an example using DifferentialEquations for completeness and this is now ready for review, thanks!

File: Project.toml [hunk 1]
```diff
@@ -3,7 +3,10 @@ uuid = "41889b7c-c35c-4f16-abd7-94b299d9fd29"
 version = "0.0.1"
 
 [deps]
+DifferentialEquations = "0c46a032-eb83-5123-abaf-570d42b7fbaa"
+OceanBioME = "a49af516-9db8-4be4-be45-1dad61c5a376"
 Oceananigans = "9e8cae18-63c1-5223-a75c-80ca9d6e9a09"
+Plots = "91a5bcdd-55d7-5caf-9e0b-520d859cae80"
 
 [compat]
```
[the remark this answers is not in the document]
Comment (nanophyto) [reply Project.toml:11]: OceanBioME forces a compat on Oceananigans.jl <0.92.0 so we don't need to add an oceananigans compat if we go this route

**Review by nanophyto — APPROVED** [review]
looks great to me, I was running into a bug yesterday (which is now fixed). I initially thought the bug was due to compat issues, but it would be good to consider nonetheless.

**Comment by nanophyto** [comment]
One other quick comment/question. In a previous version you were using notebooks, any reason you moved to `.jl` for the examples? Plot rendering within notebooks, and seeing output on GitHub without needing to run the scripts is a nice feature.

**Comment by radka-j** [comment]
> One other quick comment/question. In a previous version you were using notebooks, any reason you moved to `.jl` for the examples? Plot rendering within notebooks, and seeing output on GitHub without needing to run the scripts is a nice feature.

Good question! I mostly removed it because of the additional Jupyter dependancy (there's also the extra step of adding the IJulia kernel) but then again, that just needs to be captured in the README. Some people also advise against pushing notebooks including outputs like images because it makes the diffs bigger but personally I agree with you that I like seeing the outputs in examples when I view them on GitHub. I was also thinking of having a look at the Julia notebooks (Pluto) just to keep it strictly in the Julia ecosystem but am happy to use Jupyter if you are.

**Comment by radka-j** [comment]
> > ## For reviewers
> > Some questions:
> > 
> > * should the `run_box_model() function take in other input arguments?
> 
> It might make sense to pass the PAR function (and other aux functions) at this level? And then define the example PAR as the default?
> 
> > * we use the same PAR function definition in a few places, should this be in the library somewhere?
> 
> Yes, I think it makes sense to include the PAR function as part of the `Library`

what do you think is the best place for it? Should we have a PAR module and should it be called PAR or something else?

File: Project.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (radka-j) [reply Project.toml:11]: Good point! Yes lets do that, I think it's a fairly recent addition.

**Comment by nanophyto** [comment]
> > > ## For reviewers
> > > Some questions:
> > > 
> > > * should the `run_box_model() function take in other input arguments?
> > 
> > 
> > It might make sense to pass the PAR function (and other aux functions) at this level? And then define the example PAR as the default?
> > > * we use the same PAR function definition in a few places, should this be in the library somewhere?
> > 
> > 
> > Yes, I think it makes sense to include the PAR function as part of the `Library`
> 
> what do you think is the best place for it? Should we have a PAR module and should it be called PAR or something else?

To be consistent with the DARWIN model we could call it "non-spectral light", and rename the current "light" module to "photosynthesis"? 

https://darwin3.readthedocs.io/en/latest/phys_pkgs/darwin_light.html

The other option is to call the module simply "light" if  "non-spectral light" is too long.

**Comment by radka-j** [comment]
I  like `light` and `photosynthesis` if you're happy with that.

**Comment by nanophyto** [comment]
> > One other quick comment/question. In a previous version you were using notebooks, any reason you moved to `.jl` for the examples? Plot rendering within notebooks, and seeing output on GitHub without needing to run the scripts is a nice feature.
> 
> Good question! I mostly removed it because of the additional Jupyter dependancy (there's also the extra step of adding the IJulia kernel) but then again, that just needs to be captured in the README. Some people also advise against pushing notebooks including outputs like images because it makes the diffs bigger but personally I agree with you that I like seeing the outputs in examples when I view them on GitHub. I was also thinking of having a look at the Julia notebooks (Pluto) just to keep it strictly in the Julia ecosystem but am happy to use Jupyter if you are.

I've found Pluto difficult to use/install in the past, so I would prefer sticking to Jupyter. The dependency thing is true, but presumably facing breaking changes down the line will be a less a concern than with e.g. Oceananigans.jl or DifferentialEquations.jl. I'm agnostic in regards to larger diffs, but we can discuss it with @ljwolf this afternoon.

**Comment by radka-j** [comment]
Good to know about Pluto, I've never tried it. I am happy to use Jupyter and include the outputs in the committed notebooks, it's a workflow that I find useful. Lets double check with Levi as you suggest and take it from there.

**Comment by radka-j** [comment]
@nanophyto could you also have a look at this [commit](https://github.com/agate-model/Agate.jl/pull/41/commits/b66ae653660578452d42a5a21cc3121628500349) adding the PAR functions and let me know whether it looks alright?

File: src/Library/light.jl [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+
+"""
+    PAR_box(t, z) -> Float
+
+Time-dependent PAR at depth z.
+"""
+function PAR_box(t; z=-10)
+    return PAR⁰(t) * exp(0.2z)
+end
```
Comment (nanophyto) [inline src/Library/light.jl]: I would be tempted to merge the two functions into one with a default z value of 0 and call it `cyclical_PAR`

**Review by nanophyto — COMMENTED** [review]
Some minor comments in regards to the PAR function naming and formatting.
There is scope to tidy the PAR function further (e.g. ensure it is centred around the 21st of June, and add a hemisphere argument) but that is better suited for a separate PR.

File: src/Library/light.jl [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+const year = years = 365day
+
+
+"""
+    PAR⁰(t) -> Float
+
+Time-dependent surface-level PAR.
+"""
+function PAR⁰(t)
```
[the remark this answers is not in the document]
Comment (radka-j) [reply src/Library/light.jl:17]: I like that

**Comment by nanophyto** [comment]
> @nanophyto are you happy for me to merge this and create separate issues for cleaning up the PAR function (I already renamed it to cyclical_PAR as suggested) and for creating example notebooks just so as not to overload this PR?

yes, go for it

**Pull request merged**